tutorials/financial-tick-data/ingest-real-time-websocket-data.md

```diff
@@ -0,0 +1,438 @@
+websocket object and establish connection.
+
+## Set up a new Python environment
+
+Create a new Python virtual environment for this project and activate it. All
+the packages you need to complete for this tutorial are installed in this environment.
+
+<Procedure>
+
+### Setting up a new Python environment
+
+1.  Create and activate a Python virtual environment:
+
+    ```bash
+    virtualenv env
+    source env/bin/activate
+    ```
+
+1.  Install the Twelve Data Python
+    [wrapper library][twelve-wrapper]
+    with websocket support. This library makes it easy to make requests to the
+    API and maintain a stable websocket connection.
+
+    ```bash
+    pip install twelvedata websocket-client
+    ```
+
+1.  Install [Psycopg2][psycopg2] so that you can connect the
+    TimescaleDB from your Python script:
+
+    ```bash
+    pip install psycopg2-binary
+    ```
+
+</Procedure>
+
+## Create the websocket connection
+
+A persistent connection between your computer and the websocket server is used
+to receive data for as long as the connection is maintained. You need to pass
+two arguments to create a websocket object and establish connection.
+
+### Websocket arguments
+
+*   `on_event`
+
+    This argument needs to be a function that is invoked whenever there's a
+    new data record is received from the websocket:
+
+    ```python
+    def on_event(event):
+        print(event) # prints out the data record (dictionary)
+    ```
+
+    This is where you want to implement the ingestion logic so whenever
+    there's new data available you insert it into the database.
+
+*   `symbols`
+
+    This argument needs to be a list of stock ticker symbols (for example,
+    `MSFT`) or crypto trading pairs (for example, `BTC/USD`). When using a
+    websocket connection you always need to subscribe to the events you want to
+    receive. You can do this by using the `symbols` argument or if your
+    connection is already created you can also use the `subscribe()` function to
+    get data for additional symbols.
+
+<Procedure>
+
+### Connecting to the websocket server
+
+1.  Create a new Python file called `websocket_test.py` and connect to the
+    Twelve Data servers using the `<YOUR_API_KEY>`:
+
+    ```python
+       import time
+       from twelvedata import TDClient
+
+        messages_history = []
+
+        def on_event(event):
+         print(event) # prints out the data record (dictionary)
+         messages_history.append(event)
+
+       td = TDClient(apikey="<YOUR_API_KEY>")
+       ws = td.websocket(symbols=["BTC/USD", "ETH/USD"], on_event=on_event)
+       ws.subscribe(['ETH/BTC', 'AAPL'])
+       ws.connect()
+       while True:
+       print('messages received: ', len(messages_history))
+       ws.heartbeat()
+       time.sleep(10)
+    ```
+
+1.  Now run the Python script:
+
+    ```bash
+    python websocket_test.py
+    ```
+
+1.  After running the script, you immediately get a response from the server
```

Comment:
Isn't this a result statement? Can we combine it with the previous step?

Reply:
Done!!! these were two-part results.